src/lib/ext-activitylog.js

```diff
@@ -3,26 +3,182 @@ import { save } from './save-load.js';
+    this.checkboxTemplate = document.querySelector('#filterCheckbox').content;
 
-    this.saveLogBtn.addEventListener('click', function () {
-      this.dispatchEvent(new CustomEvent('savelog'));
-    });
+    this.extCheckboxList = document.querySelector('.checkbox-list.extension');
+    this.extFilterBtn = document.querySelector('.toggle-btn.extension');
+
+    this.viewTypeCheckboxList = document.querySelector(
+      '.checkbox-list.viewtype'
+    );
+    this.viewTypeFilterBtn = document.querySelector('.toggle-btn.viewtype');
+
+    this.extFilterBtn.addEventListener('click', this);
+    this.extCheckboxList.addEventListener('change', this);
+    this.viewTypeFilterBtn.addEventListener('click', this);
+    this.viewTypeCheckboxList.addEventListener('change', this);
+    this.saveLogBtn.addEventListener('click', this);
+  }
+
+  handleEvent(event) {
+    if (event.type === 'click') {
+      switch (event.currentTarget) {
+        case this.saveLogBtn:
+          this.saveLogBtn.dispatchEvent(new CustomEvent('savelog'));
+          break;
+        case this.extFilterBtn:
+          this.toggleFilterListDisplay(this.extFilterBtn, this.extCheckboxList);
+          break;
+        case this.viewTypeFilterBtn:
+          this.toggleFilterListDisplay(
+            this.viewTypeFilterBtn,
+            this.viewTypeCheckboxList
+          );
+          break;
+        default:
+          throw new Error(`wrong event target - ${event.target.tagName}`);
+      }
+    } else if (event.type === 'change') {
+      this.handleChangeOnCheckbox(event);
+    } else {
+      throw new Error(`wrong event type - ${event.type}`);
+    }
+  }
+
+  toggleFilterListDisplay(toggleBtn, checkboxList) {
+    checkboxList.hidden = !checkboxList.hidden;
+    this.toggleFilterBtnIcon(toggleBtn, checkboxList.hidden);
+  }
+
+  toggleFilterBtnIcon(element, isListHidden) {
+    isListHidden
+      ? element
+          .querySelector('.arrow-down')
+          .classList.replace('arrow-down', 'arrow-right')
+      : element
+          .querySelector('.arrow-right')
+          .classList.replace('arrow-right', 'arrow-down');
+  }
+
+  handleChangeOnCheckbox(event) {
+    let target = null;
+    switch (event.currentTarget) {
+      case this.extCheckboxList:
+        target = 'id';
+        break;
+      case this.viewTypeCheckboxList:
+        target = 'viewType';
+        break;
+      default:
+        break;
+    }
+
+    if (target) {
+      const filterObj = {
+        filterParam: { value: event.target.closest('input').value, target },
+        checked: event.target.closest('input').checked,
+      };
+
+      const filterEvent = new CustomEvent('filter', { detail: filterObj });
+      event.currentTarget.dispatchEvent(filterEvent);
+    }
+  }
+
+  handleNewLog(object) {
+    this.logView.addNewRow(object);
+    this.updateFilterCheckboxes(object.log);
+  }
+
+  updateFilterCheckboxes(log) {
+    for (const key of Object.keys(log)) {
```

Comment:
Should I keep the IDs in a `Set` in model ? For example, making an `filterOptions{ ids:[], viewTypes:[], types:[] }` object.

Reply:
If you create a web component, it would probably be more natural to maintain the set within the component (because it is an implementation detail). In practice the contents of this internal set would be mirrored to the model. Despite this, it still makes sense to maintain the set in the component, because it makes it easier to refactor / maintain the code.

The model should be read-only, and only be writable through controlled methods. Otherwise the model could potentially be updated and break other parts of the code that didn't expect it, and it becomes very difficult to reason about the state of the model.

Reply:
Yeah, now with a web component, every component manages its own checkboxes. Life become easier :-)